### ISSUE TITLE
`forge test` doesn't utilize all available threads for fuzzing/invariants

### ISSUE THREAD

**Issue by Philogy** [foundry-rs/foundry#8898]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (03ea54c 2024-09-18T16:31:08.207328652Z)

### What command(s) is the bug in?

forge test

### Operating System

None

### Describe the bug

Forge testing is multi-threaded but only on a per test level, for instance if I have 15 different fuzz runs and 224 available threads foundry will assign 15/224 threads with one test each, however some tests take much longer to run than others. Furthermore individual runs of the same test are completely independent. I'd expect different runs to get scheduled to different threads but foundry doesn't seem to do that, leaving a lot of performance on the table.

**Comment by grandizzy** [comment]
good point, we're currently limited by proptest test runner but should consider changing this post 1.0 release

**Comment by Philogy** [comment]
> good point, we're currently limited by proptest test runner but should consider changing this post 1.0 release

This would be an amazing feature to have, would mean a 8-224x speedup depending on the machine I run the tests on

**Comment by Philogy** [comment]
Could there be a temporary workaround where foundry duplicates a given test or invariant for you by the number of available threads and then does `total_runs / N` runs on each?

**Comment by zerosnacks** [comment]
You are correct, we use `T-breaking` as an internal tag to indicate it would likely require making a _breaking change_  and we should decide whether to include it before or after `1.0`. Moving away from `proptest` would be a huge change but indeed should not be experienced as breaking by the end-user.

**Comment by Philogy** [comment]
I see, good to know thx

**Comment by 0xalpharush** [comment]
I believe implementing this is unblocked by https://github.com/foundry-rs/foundry/pull/10190

I anticipate that there is a diminishing return for adding threads and, at some point, it may be better to run multiple processes or run on multiple machines and sync the corpuses intermittently...

**Comment by figtracer** [comment]
retested on current master: a 32-run invariant campaign passes with `--invariant-workers 2`. closing as completed.
